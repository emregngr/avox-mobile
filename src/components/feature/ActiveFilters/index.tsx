import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { memo, useCallback, useMemo, useRef } from 'react'
import { ScrollView, TouchableOpacity, View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { getActiveFilterLabel } from '@/utils/feature/getActiveFilterLabel'

interface ActiveFiltersProps {
  filters: any
  onClearAll: () => void
  onRemove: (key: string) => void
}

const AUTO_SCROLL_DELAY = 300

export const ActiveFilters = memo(({ filters, onClearAll, onRemove }: ActiveFiltersProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const getActiveFilterLabelValue = getActiveFilterLabel()

  const scrollViewRef = useRef<ScrollView>(null)

  const getFilterCount = useCallback(() => {
    let count = 0
    if (!filters) return 0
    Object?.entries(filters)?.forEach(([key, value]) => {
      if (Array?.isArray(value)) {
        count += value?.length
      } else if (value !== null && value !== undefined && value !== false) {
        count += 1
      }
    })
    return count
  }, [filters])

  const filterCount = useMemo(() => getFilterCount(), [getFilterCount])

  const filterEntries = useMemo(
    () =>
      Object.entries(filters).filter(
        ([_, value]) => value !== null && value !== undefined && value !== false,
      ),
    [filters],
  )

  const handleRemove = useCallback(
    (key: string) => {
      const elementIndex = filterEntries.findIndex(([k, _]) => k === key)
      const isInLastThree = elementIndex >= filterEntries?.length - 3

      onRemove(key)

      if (isInLastThree) {
        setTimeout(() => {
          scrollViewRef?.current?.scrollToEnd({ animated: true })
        }, AUTO_SCROLL_DELAY)
      }
    },
    [filterEntries, onRemove],
  )

  if (filterCount === 0) return null

  return (
    <View className="mr-12">
      <ScrollView
        contentContainerClassName="items-center"
        ref={scrollViewRef}
        showsHorizontalScrollIndicator={false}
        horizontal
      >
        <View className="flex-row items-center pr-4">
          <TouchableOpacity
            activeOpacity={0.7}
            className="bg-error px-3 py-1.5 rounded-full overflow-hidden mr-8 left-6"
            hitSlop={10}
            onPress={onClearAll}
            testID="active-filters-clear-button"
          >
            <ThemedText color="text-100" type="body3">
              {getLocale('clear')} ({filterCount})
            </ThemedText>
          </TouchableOpacity>

          {filterEntries.map(([key, filterValue]) => (
            <View
              className="bg-secondary-100 px-3 py-1.5 rounded-full overflow-hidden mr-2 flex-row items-center"
              key={key}
            >
              <ThemedText color="text-100" type="body3">
                {getActiveFilterLabelValue(key, filterValue)}
              </ThemedText>
              <TouchableOpacity
                activeOpacity={0.7}
                className="bg-background-quaternary rounded-full overflow-hidden ml-2"
                hitSlop={10}
                onPress={() => handleRemove(key)}
                testID={`active-filters-remove-${key}`}
              >
                <MaterialCommunityIcons color={colors?.onPrimary100} name="close" size={14} />
              </TouchableOpacity>
            </View>
          ))}
        </View>
      </ScrollView>
    </View>
  )
})
