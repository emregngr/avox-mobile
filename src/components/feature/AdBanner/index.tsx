import React, { memo, useState } from 'react'
import { View } from 'react-native'
import { BannerAd, BannerAdSize } from 'react-native-google-mobile-ads'
import { useSafeAreaInsets } from 'react-native-safe-area-context'

import { isProduction } from '@/config/env/environment'
import { AD_KEYWORDS } from '@/constants/adKeywords'
import { cn } from '@/utils/common/cn'

interface AdBannerProps {
  adUnitId: string
}

export const AdBanner = memo(({ adUnitId }: AdBannerProps) => {
  const { bottom } = useSafeAreaInsets()

  const [isAdLoaded, setIsAdLoaded] = useState<boolean>(false)

  const requestOptions = {
    keywords: AD_KEYWORDS,
    requestNonPersonalizedAdsOnly: true,
  }

  const handleAdLoaded = () => {
    setIsAdLoaded(true)
  }

  const handleAdFailedToLoad = () => {
    setIsAdLoaded(false)
  }

  return !__DEV__ && isProduction() ? (
    <View
      className={cn('bg-background-primary', isAdLoaded ? 'h-auto' : 'h-0')}
      style={{ paddingBottom: bottom }}
      testID="ad-banner-container"
    >
      <BannerAd
        onAdFailedToLoad={handleAdFailedToLoad}
        onAdLoaded={handleAdLoaded}
        requestOptions={requestOptions}
        size={BannerAdSize.ANCHORED_ADAPTIVE_BANNER}
        unitId={adUnitId}
      />
    </View>
  ) : null
})

AdBanner.displayName = 'AdBanner'
