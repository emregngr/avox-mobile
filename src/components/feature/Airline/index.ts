export * from './AirlineCard'
export * from './AirlineDetail'
export * from './AirlineFavoriteItemCard'
export * from './AirlinesTab'
export * from './SkeletonAirlineCard'
