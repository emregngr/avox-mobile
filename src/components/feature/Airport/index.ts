export * from './AirportCard'
export * from './AirportCardSkeleton'
export * from './AirportDetail'
export * from './AirportsTab'
