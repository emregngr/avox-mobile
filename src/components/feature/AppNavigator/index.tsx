import { Stack } from 'expo-router'
import { useMemo } from 'react'

import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

interface AppNavigatorProps {
  isAppReady: boolean | null
}

export const AppNavigator = ({ isAppReady }: AppNavigatorProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const stackScreenOptions = useMemo(
    () => ({
      animation: 'slide_from_bottom' as const,
      contentStyle: { backgroundColor: colors?.background?.blur },
      headerShown: false,
    }),
    [colors],
  )

  if (!isAppReady) {
    return (
      <Stack initialRouteName="index" screenOptions={stackScreenOptions}>
        <Stack.Screen name="index" />
      </Stack>
    )
  }

  return (
    <Stack initialRouteName="index" screenOptions={stackScreenOptions}>
      <Stack.Screen name="index" />
      <Stack.Screen name="(auth)" />
      <Stack.Screen name="(tabs)" />
      <Stack.Screen name="(onboarding)" />
      <Stack.Screen name="(maintenance)" />
      <Stack.Screen name="(force-update)" />
      <Stack.Screen name="(account)" />
      <Stack.Screen name="(settings)" />
      <Stack.Screen name="(airline)" />
      <Stack.Screen name="(airport)" />
      <Stack.Screen name="(breaking-news)" />
      <Stack.Screen name="(destination)" />
      <Stack.Screen name="(airplane)" />
      <Stack.Screen name="token-expire" />
      <Stack.Screen name="(web-view-modal)" options={{ presentation: 'modal' }} />
      <Stack.Screen name="(image-modal)" options={{ presentation: 'modal' }} />
    </Stack>
  )
}
