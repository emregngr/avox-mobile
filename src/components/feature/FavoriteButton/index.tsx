import { Ionicons } from '@expo/vector-icons'
import React, { memo, useMemo } from 'react'
import { ActivityIndicator, TouchableOpacity } from 'react-native'

import { useFavoriteToggle } from '@/hooks/services/useFavoriteToggle'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { FavoriteItem } from '@/types/feature/favorite'

const FavoriteButtonComponent = ({ id, type }: FavoriteItem) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { handleFavoritePress, isFavorite, isPending } = useFavoriteToggle({ id, type })

  return (
    <TouchableOpacity
      activeOpacity={0.7}
      className="bg-background-quaternary p-2 rounded-full overflow-hidden absolute top-2 right-2"
      disabled={isPending}
      hitSlop={20}
      onPress={handleFavoritePress}
    >
      {isPending ? (
        <ActivityIndicator color={colors?.tertiary100} size="small" />
      ) : (
        <Ionicons
          color={isFavorite ? colors?.tertiary100 : colors?.onPrimary100}
          name={isFavorite ? 'heart' : 'heart-outline'}
          size={20}
        />
      )}
    </TouchableOpacity>
  )
}

export const FavoriteButton = memo(FavoriteButtonComponent)
