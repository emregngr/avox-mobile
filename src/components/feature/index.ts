export * from './ActiveFilters'
export * from './Airline'
export * from './Airport'
export * from './AppNavigator'
export * from './FavoriteButton'
export * from './Favorites'
export * from './FilterModal'
export * from './Home'
