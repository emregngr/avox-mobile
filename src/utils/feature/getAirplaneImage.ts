const imageFiles = {
  airbus_a320neo: require('@/assets/images/aircraft/airbus_a320neo.webp'),
  'airbus_a350-900': require('@/assets/images/aircraft/airbus_a350-900.webp'),
  'airbus_a380-800': require('@/assets/images/aircraft/airbus_a380-800.webp'),
  'antonov_an-148': require('@/assets/images/aircraft/antonov_an-148.webp'),
  'atr_72-600': require('@/assets/images/aircraft/atr_72-600.webp'),
  'boeing_737-800': require('@/assets/images/aircraft/boeing_737-800.webp'),
  'boeing_747-8': require('@/assets/images/aircraft/boeing_747-8.webp'),
  'boeing_777-300ER': require('@/assets/images/aircraft/boeing_777-300ER.webp'),
  'boeing_787-9': require('@/assets/images/aircraft/boeing_787-9.webp'),
  bombardier_q400: require('@/assets/images/aircraft/bombardier_q400.webp'),
  cessna_citation_cj4: require('@/assets/images/aircraft/cessna_citation_cj4.webp'),
  comac_c919: require('@/assets/images/aircraft/comac_c919.webp'),
  'douglas_md-11f': require('@/assets/images/aircraft/douglas_md-11f.webp'),
  embraer_e190: require('@/assets/images/aircraft/embraer_e190.webp'),
  fokker_100: require('@/assets/images/aircraft/fokker_100.webp'),
  'ilyushin_II-76': require('@/assets/images/aircraft/ilyushin_II-76.webp'),
}

const imageList = [
  'airbus_a320neo',
  'airbus_a350-900',
  'airbus_a380-800',
  'boeing_737-800',
  'boeing_747-8',
  'boeing_777-300ER',
  'boeing_787-9',
  'embraer_e190',
  'bombardier_q400',
  'atr_72-600',
  'ilyushin_II-76',
  'antonov_an-148',
  'cessna_citation_cj4',
  'comac_c919',
  'douglas_md-11f',
  'fokker_100',
] as const

type ImageType = (typeof imageList)[number]

const specificMappings: { [key: string]: ImageType } = {
  'airbus a319': 'airbus_a320neo',
  'airbus a320neo': 'airbus_a320neo',
  'airbus a321': 'airbus_a320neo',
  'airbus a350-1000': 'airbus_a350-900',
  'airbus a350-900': 'airbus_a350-900',
  'airbus a380-800': 'airbus_a380-800',
  'atr 72-600': 'atr_72-600',
  'boeing 737 max': 'boeing_737-800',
  'boeing 737-800': 'boeing_737-800',
  'boeing 737-900': 'boeing_737-800',
  'boeing 747-400': 'boeing_747-8',
  'boeing 747-400er': 'boeing_747-8',
  'boeing 747-400f': 'boeing_747-8',
  'boeing 747-8': 'boeing_747-8',
  'boeing 747-8f': 'boeing_747-8',
  'boeing 747-8i': 'boeing_747-8',
  'boeing 777-200': 'boeing_777-300ER',
  'boeing 777-300er': 'boeing_777-300ER',
  'boeing 787-10': 'boeing_787-9',
  'boeing 787-8': 'boeing_787-9',
  'boeing 787-9': 'boeing_787-9',
  'bombardier q400': 'bombardier_q400',
  'embraer e190': 'embraer_e190',
  'embraer e195': 'embraer_e190',
}

const familyMappings: { [key: string]: ImageType } = {
  ATR: 'atr_72-600',
  'Airbus A220': 'airbus_a320neo',
  'Airbus A300': 'airbus_a320neo',
  'Airbus A310': 'airbus_a320neo',
  'Airbus A318': 'airbus_a320neo',
  'Airbus A320': 'airbus_a320neo',
  'Airbus A330': 'airbus_a350-900',
  'Airbus A340': 'airbus_a350-900',
  'Airbus A350': 'airbus_a350-900',
  'Airbus A380': 'airbus_a380-800',
  Antonov: 'antonov_an-148',
  Beechcraft: 'cessna_citation_cj4',
  'Boeing 717': 'boeing_737-800',
  'Boeing 737': 'boeing_737-800',
  'Boeing 747': 'boeing_747-8',
  'Boeing 757': 'boeing_777-300ER',
  'Boeing 767': 'boeing_777-300ER',
  'Boeing 777': 'boeing_777-300ER',
  'Boeing 787': 'boeing_787-9',
  'Bombardier CRJ': 'embraer_e190',
  'Bombardier Q': 'bombardier_q400',
  Cessna: 'cessna_citation_cj4',
  Comac: 'comac_c919',
  'De Havilland': 'bombardier_q400',
  'Douglas MD': 'douglas_md-11f',
  'Embraer E': 'embraer_e190',
  Falcon: 'cessna_citation_cj4',
  Fokker: 'fokker_100',
  Ilyushin: 'ilyushin_II-76',
  Jetstream: 'atr_72-600',
  Saab: 'atr_72-600',
  Sukhoi: 'boeing_737-800',
  Tupolev: 'ilyushin_II-76',
}

const getAirplaneImageKey = (aircraftType: string): ImageType => {
  const defaultImage: ImageType = 'airbus_a320neo'
  let imageName: ImageType = defaultImage

  const normalizedForFile = aircraftType?.toLowerCase()?.replace(/[\s-]/g, '_')
  const normalizedLower = aircraftType?.toLowerCase()

  if (imageList?.some(img => img === normalizedForFile)) {
    imageName = normalizedForFile as ImageType
  } else if (specificMappings?.[normalizedLower]) {
    imageName = specificMappings?.[normalizedLower]
  } else if (normalizedLower?.endsWith('f') && !normalizedLower?.includes('embraer')) {
    imageName = 'douglas_md-11f'
  } else {
    for (const [family, image] of Object?.entries(familyMappings)) {
      if (normalizedLower?.includes(family?.toLowerCase())) {
        imageName = image
        break
      }
    }
  }

  return imageName
}

const getAirplaneImageSource = (imageKey: ImageType): string => imageFiles?.[imageKey]

export { getAirplaneImageKey, getAirplaneImageSource, ImageType }
