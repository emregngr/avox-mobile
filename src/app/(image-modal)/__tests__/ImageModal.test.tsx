import { fireEvent, render } from '@testing-library/react-native'
import { router, useLocalSearchParams } from 'expo-router'
import type React from 'react'
import type { ReactNode } from 'react'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import ImageModal from '@/app/(image-modal)/image-modal'
import useThemeStore from '@/store/theme'
import { getAirplaneImageSource } from '@/utils/feature/getAirplaneImage'

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

jest.mock('@/components/common', () => {
  const { View, Text } = require('react-native')

  return {
    SafeLayout: ({ children }: { children: ReactNode }) => (
      <View testID="safe-layout">{children}</View>
    ),

    ThemedText: ({ children, ...props }: { children: string }) => (
      <Text {...props}>{children}</Text>
    ),
  }
})

jest.mock('@/assets/icons/close', () => 'Close')

jest.mock('@likashefqet/react-native-image-zoom', () => {
  const { View } = require('react-native')

  return {
    Zoomable: ({ children }: { children: ReactNode }) => <View testID="zoomable">{children}</View>,
  }
})

const mockedUseLocalSearchParams = useLocalSearchParams as jest.MockedFunction<
  typeof useLocalSearchParams
>

jest.mock('@/utils/feature/getAirplaneImage')

const mockedGetAirplaneImageSource = getAirplaneImageSource as jest.MockedFunction<
  typeof getAirplaneImageSource
>

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({ selectedTheme: 'light' })
})

describe('ImageModal', () => {
  it('should render the title and image correctly', () => {
    const mockedTitle = 'F-16 Fighting Falcon'
    const mockedImageKey = 'f16'
    const mockedImageSource = { uri: 'path/to/f16.jpg' }

    mockedUseLocalSearchParams.mockReturnValue({
      title: mockedTitle,
      selectedImageKey: mockedImageKey,
    })

    mockedGetAirplaneImageSource.mockReturnValue(mockedImageSource as any)

    const { getByText, getByTestId } = renderWithSafeAreaProvider(<ImageModal />)

    expect(getByText(mockedTitle)).toBeTruthy()

    expect(mockedGetAirplaneImageSource).toHaveBeenCalledWith(mockedImageKey)
    const image = getByTestId('mocked-image')
    expect(image.props.source).toEqual(mockedImageSource)
  })

  it('should call router.back when the close button is pressed', () => {
    mockedUseLocalSearchParams.mockReturnValue({
      title: 'Any Title',
      selectedImageKey: 'any-key',
    })
    mockedGetAirplaneImageSource.mockReturnValue({ uri: 'path/to/any.jpg' } as any)

    const { getByTestId } = renderWithSafeAreaProvider(<ImageModal />)

    const closeButton = getByTestId('close-button')

    fireEvent.press(closeButton)

    expect(router.back).toHaveBeenCalledTimes(1)
  })

  it('should render gracefully with a null image source', () => {
    mockedUseLocalSearchParams.mockReturnValue({
      title: 'Title Without Image',
      selectedImageKey: undefined,
    } as any)

    const { getByText, getByTestId } = renderWithSafeAreaProvider(<ImageModal />)

    expect(getByText('Title Without Image')).toBeTruthy()
    const image = getByTestId('mocked-image')
    expect(image.props.source).toBeNull()

    expect(mockedGetAirplaneImageSource).not.toHaveBeenCalled()
  })
})

describe('ImageModal Screen Snapshot', () => {
  it('should render the ImageModal Screen successfully', () => {
    mockedUseLocalSearchParams.mockReturnValue({
      title: 'Any Title',
      selectedImageKey: 'any-key',
    })

    mockedGetAirplaneImageSource.mockReturnValue({ uri: 'path/to/any.jpg' } as any)

    const { toJSON } = renderWithSafeAreaProvider(<ImageModal />)

    expect(toJSON()).toMatchSnapshot()
  })
})
