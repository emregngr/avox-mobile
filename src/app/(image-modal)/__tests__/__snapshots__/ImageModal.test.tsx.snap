// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ImageModal Screen Snapshot should render the ImageModal Screen successfully 1`] = `
<View
  testID="safe-layout"
>
  <View
    className="flex-1 bg-background-primary"
  >
    <View
      className="flex-row items-center justify-between p-4 z-10 bg-transparent"
    >
      <View
        className="flex-1 ml-10 mr-4"
      >
        <Text
          center={true}
          color="text-100"
          ellipsizeMode="tail"
          numberOfLines={2}
          type="h3"
        >
          Any Title
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={20}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="close-button"
      >
        <Close
          height={24}
          primaryColor="#D1D1D6"
          secondaryColor="#000000"
          width={24}
        />
      </View>
    </View>
    <View
      testID="zoomable"
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          cachePolicy="memory-disk"
          contentFit="contain"
          source={
            {
              "uri": "path/to/any.jpg",
            }
          }
          style={
            {
              "height": "100%",
              "width": "100%",
            }
          }
          testID="mocked-image"
          transition={0}
        />
      </View>
    </View>
  </View>
</View>
`;
